chain/network/src/routing.rs

```diff
@@ -60,68 +47,99 @@ pub struct Edge {
     peer0: PeerId,
     peer1: PeerId,
     /// Nonce to keep tracking of the last update on this edge.
-    nonce: u64,
-    /// Specify status of the edge in the last update.
-    edge_type: EdgeType,
-    /// Signature from parties validating the edge.
-    /// If the edge_type is Removed, only one signature is required.
-    signature0: Option<Signature>,
-    signature1: Option<Signature>,
+    /// It must be even
+    pub nonce: u64,
+    /// Signature from parties validating the edge. These are signature of the added edge.
+    signature0: Signature,
+    signature1: Signature,
+    /// Info necessary to declare an edge as removed.
+    /// The bool says which party is removing the edge: false for Peer0, true for Peer1
+    /// The signature from the party removing the edge.
+    removal_info: Option<(bool, Signature)>,
 }
 
 impl Edge {
+    /// Create an addition edge.
     pub fn new(
         peer0: PeerId,
         peer1: PeerId,
         nonce: u64,
-        edge_type: EdgeType,
-        signature0: Option<Signature>,
-        signature1: Option<Signature>,
+        signature0: Signature,
+        signature1: Signature,
     ) -> Self {
         let (peer0, signature0, peer1, signature1) = if peer0 < peer1 {
             (peer0, signature0, peer1, signature1)
         } else {
             (peer1, signature1, peer0, signature0)
         };
 
-        Self { peer0, peer1, nonce, edge_type, signature0, signature1 }
+        Self { peer0, peer1, nonce, signature0, signature1, removal_info: None }
+    }
+
+    /// Create the remove edge change from an added edge change.
+    pub fn remove_edge(&self, me: PeerId, sk: SecretKey) -> Self {
+        assert_eq!(self.edge_type(), EdgeType::Added);
+        let mut edge = self.clone();
+        edge.nonce += 1;
+        let me = edge.peer0 == me;
+        let hash = edge.hash();
+        let signature = sk.sign(hash.as_ref());
+        edge.removal_info = Some((me, signature));
+        edge
     }
 
-    fn hash(peer0: &PeerId, peer1: &PeerId, nonce: u64, edge_type: &EdgeType) -> CryptoHash {
+    fn build_hash(peer0: &PeerId, peer1: &PeerId, nonce: u64) -> CryptoHash {
         let mut buffer = Vec::<u8>::new();
         let peer0: Vec<u8> = peer0.clone().into();
         buffer.extend_from_slice(peer0.as_slice());
         let peer1: Vec<u8> = peer1.clone().into();
         buffer.extend_from_slice(peer1.as_slice());
         buffer.write_u64::<LittleEndian>(nonce).unwrap();
-        buffer.push(edge_type.value());
         hash(buffer.as_slice())
     }
 
+    fn hash(&self) -> CryptoHash {
+        Edge::build_hash(&self.peer0, &self.peer1, self.nonce)
+    }
+
+    fn prev_hash(&self) -> CryptoHash {
+        Edge::build_hash(&self.peer0, &self.peer1, self.nonce - 1)
```

Comment:
If `self.nonce` is 0 this will crash

Reply:
It is only called once in the code, and I check that `self.nonce > 0` before doing that.